src/java/org/apache/cassandra/db/Columns.java

```diff
@@ -459,42 +466,110 @@ public String toString()
+            int regularCount = 0;
+            int syntheticCount = 0;
+            long size = 0;
+
+            // Count and calculate sizes
             for (ColumnMetadata column : columns)
-                size += ByteBufferUtil.serializedSizeWithVIntLength(column.name.bytes);
-            return size;
+            {
+                if (column.isSynthetic())
+                {
+                    syntheticCount++;
+                    size += ByteBufferUtil.serializedSizeWithVIntLength(column.name.bytes);
+                    size += typeSerializer.serializedSize(column.type);
+                }
+                else
+                {
+                    regularCount++;
+                    size += ByteBufferUtil.serializedSizeWithVIntLength(column.name.bytes);
+                }
+            }
+
+            return TypeSizes.sizeofUnsignedVInt(getPackedCount(syntheticCount, regularCount)) +
+                   size;
         }
 
         public Columns deserialize(DataInputPlus in, TableMetadata metadata) throws IOException
         {
-            int length = (int)in.readUnsignedVInt();
             try (BTree.FastBuilder<ColumnMetadata> builder = BTree.fastBuilder())
             {
-                for (int i = 0; i < length; i++)
+                long packedCount = in.readUnsignedVInt() ;
+                int regularCount = (int) (packedCount & 0xFFFFF);
+                int syntheticCount = (int) (packedCount >> 20);
+
+                // First pass - regular columns
+                for (int i = 0; i < regularCount; i++)
                 {
                     ByteBuffer name = ByteBufferUtil.readWithVIntLength(in);
                     ColumnMetadata column = metadata.getColumn(name);
                     if (column == null)
                     {
-                        // If we don't find the definition, it could be we have data for a dropped column, and we shouldn't
-                        // fail deserialization because of that. So we grab a "fake" ColumnMetadata that ensure proper
-                        // deserialization. The column will be ignore later on anyway.
+                        // If we don't find the definition, it could be we have data for a dropped column
                         column = metadata.getDroppedColumn(name);
-
                         if (column == null)
                             throw new RuntimeException("Unknown column " + UTF8Type.instance.getString(name) + " during deserialization");
                     }
                     builder.add(column);
                 }
+
+                // Second pass - synthetic columns
+                for (int i = 0; i < syntheticCount; i++)
+                {
+                    ByteBuffer name = ByteBufferUtil.readWithVIntLength(in);
+                    AbstractType<?> type = typeSerializer.deserialize(in);
+
+                    // TODO generalize support for other synthetic columns
```

Comment:
Do we need to do that before merging? It seems like something that would impact serialization.

Reply:
Yes, we should resolve this.  What I'm not sure of is whether we should serialize

1. Name and type which lets the coordinator include whatever synthetic columns it wants, or
2. Create an enum of known synthetic columns and just serialize an ID

The current code is a bit of a mix of the two.

I like the simplicity of #1, but based on my experience pushing this change through I'm skeptical that it will be useful to have that level of flexibility, so maybe #2 makes more sense as an additional level of making it obvious if you try to do something unsupported.

Reply:
If I were going to argue for #1 it would be along the lines of "let's turn function calls into synthetic columns so they don't have to run coordinator side" but there is no way in hell I'm signing up for that mission and it's probably not even a performance win for anything but vectors.

Reply:
I think I'm going to leave it serializing with #1 but check at deserialization that it's an expected column.  That way we're not coding ourselves into a corner if we ever need a dynamic synthetic column name.

Reply:
> I think I'm going to leave it serializing with https://github.com/datastax/cassandra/pull/1 but check at deserialization that it's an expected column.

In that case, will we just have an assertion on the `name` result a few lines above? I just noticed that is read but not used.

Option 1 works for me.